### PR TITLE
Primary and secondary rate-limit handling

### PR DESCRIPTION
**Pull request by kfcampbell**
Note that parts of testing code were taken from our colleagues at Kiota (as noted [here](https://github.com/octokit/go-sdk/blob/rate-limit-playground/pkg/handlers/rate_limit_handler_test.go#L16)) and parts of implementation code were taken from the google/go-github library (as noted [here](https://github.com/octokit/go-sdk/blob/rate-limit-playground/pkg/handlers/rate_limit_handler.go#L128)). I'm grateful to those teams for doing excellent work before us!

This PR adds a rate limit handler to the go-sdk for primary and secondary rate limits. When it detects such a rate limit, it sleeps until the limit has expired, then retries the request. This has been validated using a local GitHub test environment in addition to unit test coverage. In the future, it would be cool to build better retry logic that builds in exponential backoff, jitter, throttling, etc.

It also greatly simplifies the abstractions for API client instantiation. The values I focused on were: 
- Minimal chaining of objects (I was annoyed at constructing a handler, to pass to an array of middleware, to pass to an an HTTP client, to pass to an adapter, to pass to an API client)
- Functional initialization pattern
- Sensible defaults when not specified

The current pattern looks like this: 

```go
client, err := pkg.NewApiClient(
	pkg.WithUserAgent("my-user-agent"),
	pkg.WithRequestTimeout(5*time.Second),
	pkg.WithBaseUrl("https://api.github.com"),
	pkg.WithAuthorizationToken(os.Getenv("GITHUB_TOKEN")),
)

// equally valid:
//client, err := pkg.NewApiClient()
if err != nil {
	log.Fatalf("error creating client: %v", err)
}
```

### CODE REVIEW

File: pkg/authentication/example_test.go [hunk 1]
```diff
@@ -69,3 +72,146 @@ func ExampleApiClient_Octocat() {
 	//              ~~~~~~==~==~==~==~~~~~~
 	//                  :~==~==~==~==~~
 }
+
+func Example_ResponseHandlerForRateLimiting() {
```
Comment (kfcampbell) [inline pkg/authentication/example_test.go:76]: This function (currently with intentional malformed example suffix) shows using a Kiota abstraction for a response handler.

File: pkg/authentication/example_test.go [hunk 2]
```diff
@@ -69,3 +72,146 @@ func ExampleApiClient_Octocat() {
...
+	// this responseHandler function panics:
+	// panic: interface conversion: *http.Response is not serialization.Parsable: missing method GetFieldDeserializers
+	// not sure if i'm mishandling something in this totally non-implemented function or what
```
Comment (kfcampbell) [inline pkg/authentication/example_test.go:79]: I'm not sure what I'm doing wrong here (possibly related to not implementing any logic at all in the handler), but this panics each call. More investigation is required.

File: pkg/authentication/example_test.go [hunk 3]
```diff
@@ -69,3 +72,146 @@ func ExampleApiClient_Octocat() {
...
+	//                    MM MM MM MM
+	//                    MM MM MM MM
+	//                 .~~MM~MM~MM~MM~~.
+	//              ~~~~MM:~MM~~~MM~:MM~~~~
+	//             ~~~~~~==~==~~~==~==~~~~~~
+	//              ~~~~~~==~==~==~==~~~~~~
+	//                  :~==~==~==~==~~
+}
+
+func Example_HttpClientForRateLimiting() {
```
Comment (kfcampbell) [inline pkg/authentication/example_test.go:155]: This function (currently with intentional malformed example suffix) uses a Go standard library HTTP client with a RoundTripper with logic mostly cribbed from `integrations/terraform-provider-github` for rate limiting. It does not make use of the Kiota helpers.

File: pkg/headers/header_contents.go [hunk 4]
```diff
@@ -11,3 +11,7 @@ const APIVersionKey = "X-GitHub-Api-Version"
 
 // TODO(kfcampbell): get the version from the generated code somehow
 const APIVersionValue = "2022-11-28"
+
+const XRateLimitRemainingKey = "X-Ratelimit-Remaining"
+const XRateLimitResetKey = "X-Ratelimit-Reset"
+const RetryAfterKey = "Retry-After"
```
Comment (nickfloyd) [inline pkg/headers/header_contents.go:17]: Do we want to go ahead and add all of the potential rate limit headers here:
`x-ratelimit-limit`
`x-ratelimit-used`
`x-ratelimit-resource`

And perhaps expose the values as properties or logged items - to help with debugging, usage?

File: cmd/rate-limit-test/main.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+			requestConfig := &abstractions.RequestConfiguration[user.ReposRequestBuilderGetQueryParameters]{
+				QueryParameters: queryParams,
+			}
+			_, err := client.User().Repos().Get(context.Background(), requestConfig)
+			if err != nil {
+				log.Fatalf("error getting repositories: %v", err)
+				return err
+			}
+
+			// if len(repos) > 0 {
```
Comment (nickfloyd) [inline cmd/rate-limit-test/main.go:53]: Might be an unnecessary nit - comment clean up.

File: cmd/rate-limit-test/main.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
+package main
```
Comment (nickfloyd) [inline cmd/rate-limit-test/main.go:1]: I responded to you in another thread about these tests - for some further considerations
1.  Where should this be given the dependency on local
2. We should probably document the details covered in the issue you have - i.e. formalized docs and such

Comment (kfcampbell) [reply cmd/rate-limit-test/main.go]: I've removed this file and linked to a copy of it in our private issue should we need to retrieve it later.

**Review by nickfloyd — APPROVED** [review]
Looks great! ➕ For setting up the infrastructure for the local testing ground as well!

**Pull request merged**